[AArch64][GlobalISel] Improve and expand fcopysign lowering

davemgreen opened the pull request:
This alters the lowering of G_COPYSIGN to support vector types. The general idea is that we just lower it to vector operations using and/or and a mask, which are now converted to a BIF/BIT/BSP.

In the process the existing AArch64LegalizerInfo::legalizeFCopySign can be removed, replying on expanding the scalar versions to vector instead, which just needs a small adjustment to allow widening scalars to vectors.

tschuett commented:
E.g. G_ABS does not support scalars currently. With the new `moreElementsIf` we can vectorize scalars.

arsenm commented on llvm/test/CodeGen/AArch64/fcopysign.ll:
```diff
@@ -55,64 +51,109 @@ entry:
-; CHECK-NEXT:    ret
+; CHECK-SD-LABEL: copysign_v4f64:
+; CHECK-SD:       // %bb.0: // %entry
+; CHECK-SD-NEXT:    movi v4.2d, #0xffffffffffffffff
+; CHECK-SD-NEXT:    fneg v4.2d, v4.2d
+; CHECK-SD-NEXT:    bif v0.16b, v2.16b, v4.16b
+; CHECK-SD-NEXT:    bif v1.16b, v3.16b, v4.16b
+; CHECK-SD-NEXT:    ret
```
DAG looks better?

davemgreen commented:
Thanks for the reviews and sorry for the delay on this one. There is now a separate optimization to produce fneg(movi 0), which should cover the places where this was previously potentially performing worse.

davemgreen committed:
[AArch64][GlobalISel] Improve and expand fcopysign lowering

This alters the lowering of G_COPYSIGN to support vector types. The general
idea is that we just lower it to vector operations using and/or and a mask,
which are now converted to a BIF/BIT/BSP.

In the process the existing AArch64LegalizerInfo::legalizeFCopySign can be
removed, replying on expanding the scalar versions to vector instead, which
just needs a small adjustment to allow widening scalars to vectors.
